nautobot/dcim/models/devices.py

```diff
@@ -670,11 +670,17 @@ def clean(self):
 
         # Validate location
         if self.location is not None:
-            # TODO: after Location model replaced Site, which was not a hierarchical model, should we allow users to assign a Rack belongs to
-            # the parent Location or the child location of `self.location`?
-
-            if self.rack is not None and self.rack.location != self.location:
-                raise ValidationError({"rack": f'Rack "{self.rack}" does not belong to location "{self.location}".'})
+            if self.rack is not None:
+                device_location = self.location
+                # Rack's location must be a child location or the same location as that of the parent device.
+                # Location is a required field on rack.
+                rack_location = self.rack.location
+                if rack_location not in device_location.descendants(include_self=True):
```

Comment:
Might be more efficient to reverse this since typically a location will have fewer ancestors than descendants?

```suggestion
                if device_location not in rack_location.ancestors(include_self=True):
```